### PR TITLE
Add a mask rcnn example for config

### PR DESCRIPTION
**Pull request by ZwwWayne**
This PR adds comments in a complete config to explain the detailed meanings of each key.

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+To help the users have a basic idea of a complete config and the modules in a modern detection system,
+we makea brief comments on the config of Mask R-CNN using ResNet50 and FPN as the following.
+For more detailed usage and the corresponding alternative for each modules, please refer the API documentation.
+
+```python
+model=dict(
+    type='MaskRCNN',  # The name of detector
+    pretrained='torchvision://resnet50',    # The ImageNet pretrained backbone to be loaded
+    backbone=dict(  # The config of backbone
+        type='ResNet',    # The type of the backbone
```
Comment (hellock) [inline docs/config.md:57]: We may refer each component to its docstring to help users find more details. E.g., for the line `type='ResNet'`, we may refer to `https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/backbones/resnet.py#L288`.

File: docs/config.md [hunk 2]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+For more detailed usage and the corresponding alternative for each modules, please refer the API documentation.
+
+```python
+model=dict(
+    type='MaskRCNN',  # The name of detector
+    pretrained='torchvision://resnet50',    # The ImageNet pretrained backbone to be loaded
+    backbone=dict(  # The config of backbone
+        type='ResNet',    # The type of the backbone, refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/backbones/resnet.py#L288 for more details.
+        depth=50,  # The depth of backbone, usually it is 50 or 101 for ResNet and ResNext backbones.
+        num_stages=4,  # Number of stages of the backbone, 4 for ResNet and ResNext backbones.
```
Comment (hellock) [inline docs/config.md:59]: `4 for ResNet and ResNext backbones` may not be proper.

File: docs/config.md [hunk 3]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+        out_indices=(0, 1, 2, 3),  # The index of output feature maps produced in each stages
+        frozen_stages=1,  # The weights in the first 1 stage are fronzen
+        norm_cfg=dict(  # The config of normalization layers.
+            type='BN',  # Type of norm layer, usually it is BN of GN
```
Comment (hellock) [inline docs/config.md:63]: typo: of -> or

File: docs/config.md [hunk 4]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            requires_grad=True),   # Whether to train the gamma and beta in BN
+        norm_eval=True,  # Whether to freeze the statistics in BN
+        style='pytorch'),  # The style of backbone, 'pytorch' means that stride 2 layers are in 3x3 conv,
+                           # 'caffe' means stride 2 layers are in 1x1 convs.
+    neck=dict(
+        type='FPN',  # The neck of detector is FPN. We also support 'NASFPN', 'PAFPN', etc. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/necks/fpn.py#L10 for more details.
+        in_channels=[256, 512, 1024, 2048],  # The input channels, usually this is consistent with the output channels of backbone
```
Comment (hellock) [inline docs/config.md:70]: `usually` may be deleted.

File: docs/config.md [hunk 5]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+        out_channels=256,  # The output channels of each level of the pyramid feature map
+        num_outs=5),  # The number of output scales
+    rpn_head=dict(
+        type='RPNHead',  # The type of RPN head is 'RPNHead', we also support 'GARPNHead', etc. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/dense_heads/rpn_head.py#L12 for more details.
+        in_channels=256,  # The input channels of each input feature map, this is consistent with the output channels of neck
+        feat_channels=256,  # Feature channels of convolutional layers in the head.
+        anchor_generator=dict(  # The config of anchor generator
+            type='AnchorGenerator',  # Most of methods use AnchorGenerator, SSD Detectors uses `SSDAnchorGenerator`. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/anchor/anchor_generator.py#L10 for more details
+            scales=[8],  # Basic scale of the anchor
+            ratios=[0.5, 1.0, 2.0],  # The ratio between width and height
```
Comment (hellock) [inline docs/config.md:80]: Confirm whether it is `width and height` and `height and width`.

File: docs/config.md [hunk 6]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            strides=[4, 8, 16, 32, 64]),  # The strides of the anchor generator. This is consistent with the FPN feature strides.
```
Comment (hellock) [inline docs/config.md:81]: We may explain the relation between anchor size and `scales`, `base_sizes` and `strides`.

File: docs/config.md [hunk 7]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+        bbox_coder=dict(  # Config of box coder to encode and decode the boxes during training and testing
+            type='DeltaXYWHBBoxCoder',  # Type of box coder. 'DeltaXYWHBBoxCoder' is applied for most of methods. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/bbox/coder/delta_xywh_bbox_coder.py#L9 for more details.
+            target_means=[0.0, 0.0, 0.0, 0.0],  # The target means used to encode and decode boxes
+            target_stds=[1.0, 1.0, 1.0, 1.0]),  # The standard variance used to encode and decode boxes
+        loss_cls=dict(  # Config of loss function for classification branch
+            type='CrossEntropyLoss',  # Type of loss for classification branch, we also support FocalLoss etc.
+            use_sigmoid=True,  # RPN usually do two-class classification, so it usually uses sigmoid function.
```
Comment (hellock) [inline docs/config.md:88]: do -> performs

File: docs/config.md [hunk 8]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            loss_weight=1.0),  # Weight of loss for classification branch.
```
Comment (hellock) [inline docs/config.md:89]: Loss weight of the classification branch.

File: docs/config.md [hunk 9]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+        anchor_generator=dict(  # The config of anchor generator
+            type='AnchorGenerator',  # Most of methods use AnchorGenerator, SSD Detectors uses `SSDAnchorGenerator`. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/anchor/anchor_generator.py#L10 for more details
+            scales=[8],  # Basic scale of the anchor
+            ratios=[0.5, 1.0, 2.0],  # The ratio between width and height
+            strides=[4, 8, 16, 32, 64]),  # The strides of the anchor generator. This is consistent with the FPN feature strides.
+        bbox_coder=dict(  # Config of box coder to encode and decode the boxes during training and testing
+            type='DeltaXYWHBBoxCoder',  # Type of box coder. 'DeltaXYWHBBoxCoder' is applied for most of methods. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/bbox/coder/delta_xywh_bbox_coder.py#L9 for more details.
+            target_means=[0.0, 0.0, 0.0, 0.0],  # The target means used to encode and decode boxes
+            target_stds=[1.0, 1.0, 1.0, 1.0]),  # The standard variance used to encode and decode boxes
+        loss_cls=dict(  # Config of loss function for classification branch
```
Comment (hellock) [inline docs/config.md:86]: the classification branch

File: docs/config.md [hunk 10]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            type='CrossEntropyLoss',  # Type of loss for classification branch, we also support FocalLoss etc.
+            use_sigmoid=True,  # RPN usually do two-class classification, so it usually uses sigmoid function.
+            loss_weight=1.0),  # Weight of loss for classification branch.
+        loss_bbox=dict(  # Config for regression branch.
+            type='L1Loss',  # Type of loss, we also support many IoU Losses and smooth L1-loss, etc.
```
Comment (hellock) [inline docs/config.md:91]: Add a ref url.

File: docs/config.md [hunk 11]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            loss_weight=1.0)),  # Weight of loss for regression branch.
+    roi_head=dict(  # RoIHead encapsulates the second stage of two-stage/cascade detectors.
+        type='StandardRoIHead',  # Type of the second stage head. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/roi_heads/standard_roi_head.py#L10 for implementation.
```
Comment (hellock) [inline docs/config.md:94]: Type of RoI head.

File: docs/config.md [hunk 12]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+        bbox_roi_extractor=dict(  # RoI feature extractor for bbox regression.
+            type='SingleRoIExtractor',  # Type of the RoI feature extractor, most of methods uses SingleRoIExtractor. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/roi_heads/roi_extractors/single_level.py#L10 for details.
+            roi_layer=dict(  # Config of RoI Layer
+                type='RoIAlign',  # Type of RoI Layer, DeformRoIPoolingPack and ModulatedDeformRoIPoolingPack are also supported. Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/ops/roi_align/roi_align.py#L79 for details.
+                out_size=7,  # The output size of feature maps.
+                sample_num=0),  # Sampling ratio when extracting the RoI features.
```
Comment (hellock) [inline docs/config.md:100]: 0 means adaptive ratio.

File: docs/config.md [hunk 13]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            out_channels=256,  # output channels of the extracted feature.
+            featmap_strides=[4, 8, 16, 32]),  # Strides of multi-scale feature maps.
```
Comment (hellock) [inline docs/config.md:102]: It should be consistent to the architecture of the backbone.

File: docs/config.md [hunk 14]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+        bbox_head=dict(  # Config of box head in the second stage.
```
Comment (hellock) [inline docs/config.md:103]: second stage -> RoIHead

File: docs/config.md [hunk 15]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            type='Shared2FCBBoxHead',  # Type of the bbox head, Refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py#L177 for implementation details.
+            in_channels=256,  # Input channels for bbox head. This is consistent with the out_channels in roi_extractor
+            fc_out_channels=1024,  # Output feature channels of FC layers.
+            roi_feat_size=7,  # Size of RoI features
+            num_classes=80,  # Number of classes for classification
+            bbox_coder=dict(  # Box coder used in the second stage.
+                type='DeltaXYWHBBoxCoder',  # Type of box coder. 'DeltaXYWHBBoxCoder' is applied for most of methods.
+                target_means=[0.0, 0.0, 0.0, 0.0],  # Means used to encode and decode box
+                target_stds=[0.1, 0.1, 0.2, 0.2]),  # Standard variance for encoding and decoding. It is smaller since the boxes are more accurate.
```
Comment (hellock) [inline docs/config.md:112]: [0.1, 0.1, 0.2, 0.2] is a conventional setting.

File: docs/config.md [hunk 16]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+    warmup_iters=500,  # The number of iterations for warmup
+    warmup_ratio=0.001,  # The ratio of the starting learning rate used for warmup
+    step=[8, 11])  # Steps to decay the learning rate
+total_epochs=12  # Total epoches to train the model
+checkpoint_config=dict(  # Config to set the checkpoint hook
+    interval=1)  # The save interval is 1
+log_config=dict(  # config to register logger hook
+    interval=50,  # Interval to print the log
+    hooks=[
+        dict(type='TextLoggerHook')])  # The logger used to record the training process.
```
Comment (hellock) [inline docs/config.md:336]: Also add TensorboardLoggerHook in the comments.

File: docs/config.md [hunk 17]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+dist_params=dict(backend='nccl')  # parameters to setup distributed training, port can also be set.
```
Comment (hellock) [inline docs/config.md:337]: the port

File: docs/config.md [hunk 18]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+log_level='INFO'  # The level of the log.
```
Comment (hellock) [inline docs/config.md:338]: The level of logging.

File: docs/config.md [hunk 19]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+load_from=None  # load models as a pre-trained model from a given path. This will not resume training.
+resume_from=None  # Resume checkpoints from a given path, the training will be resumed from the epoch when the checkpoint's is saved.
+workflow=[('train', 1)]  # Workflow for runner. This means to train the model for once.
```
Comment (hellock) [inline docs/config.md:341]: `This means to train the model for once.` makes user confusing.

File: docs/config.md [hunk 20]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+work_dir='./exps/check'  # Directory to save the model checkpoints and logs for the current experiments.
```
Comment (hellock) [inline docs/config.md:342]: Use the default work_dir.

File: docs/config.md [hunk 21]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+optimizer_config=dict(  # Config used to build the optimizer hook, refer https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/optimizer.py#L8 for implementation details.
+    grad_clip=None)  # Most of the methods do not use gradient clip
+lr_config=dict(  # Learning rate scheduler config used to register LrUpdater hook
+    policy='step',  # The policy of scheduler, also support CosineAnealing, Cyclic, etc. Refer details of supported LrUpdater from https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/lr_updater.py#L9.
+    warmup='linear',  # The warmup policy, also support `exp` and `constant`.
+    warmup_iters=500,  # The number of iterations for warmup
+    warmup_ratio=0.001,  # The ratio of the starting learning rate used for warmup
+    step=[8, 11])  # Steps to decay the learning rate
+total_epochs=12  # Total epoches to train the model
+checkpoint_config=dict(  # Config to set the checkpoint hook
```
Comment (hellock) [inline docs/config.md:331]: Refer to mmcv.

File: docs/config.md [hunk 22]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+    weight_decay=0.0001)  # Weight decay of SGD
+optimizer_config=dict(  # Config used to build the optimizer hook, refer https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/optimizer.py#L8 for implementation details.
+    grad_clip=None)  # Most of the methods do not use gradient clip
+lr_config=dict(  # Learning rate scheduler config used to register LrUpdater hook
+    policy='step',  # The policy of scheduler, also support CosineAnealing, Cyclic, etc. Refer details of supported LrUpdater from https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/lr_updater.py#L9.
+    warmup='linear',  # The warmup policy, also support `exp` and `constant`.
+    warmup_iters=500,  # The number of iterations for warmup
+    warmup_ratio=0.001,  # The ratio of the starting learning rate used for warmup
+    step=[8, 11])  # Steps to decay the learning rate
+total_epochs=12  # Total epoches to train the model
```
Comment (hellock) [inline docs/config.md:330]: epochs

File: docs/config.md [hunk 23]
```diff
@@ -43,6 +43,305 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+                        to_rgb=True),
+                    dict(type='Pad',
+                        size_divisor=32),
+                    dict(type='ImageToTensor',
+                        keys=['img']),
+                    dict(type='Collect',
+                        keys=['img'])])]))
+evaluation=dict(  # The config to build the evaluation hook, refer https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/evaluation/eval_hooks.py#L7 for more details.
+    interval=1,  # Evaluation interval
+    metric=['bbox', 'segm'])  # Metric used during evaluation
```
Comment (hellock) [inline docs/config.md:316]: Metrics

File: docs/config.md (hunk 5 above)
Comment (ZwwWayne) [reply docs/config.md:80]: it is height and width

File: docs/config.md [hunk 24]
```diff
@@ -43,6 +43,306 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+dataset_type='CocoDataset'  # Dataset type, this will be used to define the dataset
+data_root='data/coco/'  # Root path of data
+img_norm_cfg=dict(  # Image normalization config to normalize the input images
+    mean=[123.675, 116.28, 103.53],  # Mean values used to pre-training the pre-trained backbone models
+    std=[58.395, 57.12, 57.375],  # Standard variance used to pre-training the pre-trained backbone models
+    to_rgb=True)  # The channel orders of image used to pre-training the pre-trained backbone models
+train_pipeline=[  # Training pipeline
+    dict(type='LoadImageFromFile'),  # First pipeline to load images from file path
+    dict(type='LoadAnnotations',  # Second pipeline to load annotations for current image
+        with_bbox=True,  # Whether to use bounding box, True for detection
```
Comment (hellock) [inline docs/config.md:205]: The indentation seems incorrect.

File: docs/config.md [hunk 25]
```diff
@@ -43,6 +43,306 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            conv_out_channels=256,  # Output channels of the convolutional layer.
+            num_classes=80,  # Number of class to be segmented.
+            loss_mask=dict(  # Config of loss function for the mask branch.
+                type='CrossEntropyLoss',  # Type of loss used for segmentation
+                use_mask=True,  # Whether to only train the mask in the correct class.
+                loss_weight=1.0))))  # Loss weight of mask branch.
+train_cfg=dict(  # Config of training hyperparameters for rpn and rcnn
+    rpn=dict(  # Training config of rpn
+        assigner=dict(  # Config of assigner
+            type='MaxIoUAssigner',  # Type of assigner, MaxIoUAssigner is used for many common detectors.
```
Comment (hellock) [inline docs/config.md:142]: Add a URL for reference.

File: docs/config.md [hunk 26]
```diff
@@ -43,6 +43,306 @@ For `1x`/`2x`, initial learning rate decays by a factor of 10 at the 8/16th and
...
+            pos_iou_thr=0.7,  # IoU >= threshold 0.7 will be taken as positive samples
+            neg_iou_thr=0.3,   # IoU < threshold 0.3 will be taken as negative samples
+            min_pos_iou=0.3,  # The minimal IoU threshold to take boxes as positive samples
+            match_low_quality=True,  # Whether to match the boxes under low quality (see API doc for more details).
+            ignore_iof_thr=-1),  # IoF threshold for ignoring bboxes
+        sampler=dict(  # Config of positive/negative sampler
+            type='RandomSampler',  # Type of sampler, PseudoSampler and other samplers are also supported.
```
Comment (hellock) [inline docs/config.md:149]: Add a URL for reference.

**Pull request merged**